addons/product/product.py

```diff
@@ -746,7 +746,10 @@ def _get_uom_id(self, cr, uid, *args):
     def _check_uom(self, cr, uid, ids, context=None):
         for supplier_info in self.browse(cr, uid, ids, context=context):
             if supplier_info.product_uom and supplier_info.product_uom.category_id.id <> supplier_info.product_id.uom_id.category_id.id:
-                return False
```

Comment:
The real problem here is in 

    def _get_uom_id(self, cr, uid, *args):
        cr.execute('select id from product_uom order by id limit 1')
        res = cr.fetchone()
        return res and res[0] or False

That sets the same uom to all supplier_info and causes error if products have uom of different categories.  Didn't know how to fix _get_uom_id function so I needed to fix it here to have the migration done.